flink-cdc-common/src/main/java/org/apache/flink/cdc/common/data/binary/BinaryArrayData.java

```diff
@@ -0,0 +1,574 @@
+
+    private static final int BOOLEAN_ARRAY_OFFSET = UNSAFE.arrayBaseOffset(boolean[].class);
+    private static final int SHORT_ARRAY_OFFSET = UNSAFE.arrayBaseOffset(short[].class);
+    private static final int INT_ARRAY_OFFSET = UNSAFE.arrayBaseOffset(int[].class);
+    private static final int LONG_ARRAY_OFFSET = UNSAFE.arrayBaseOffset(long[].class);
+    private static final int FLOAT_ARRAY_OFFSET = UNSAFE.arrayBaseOffset(float[].class);
+    private static final int DOUBLE_ARRAY_OFFSET = UNSAFE.arrayBaseOffset(double[].class);
+
+    public static int calculateHeaderInBytes(int numFields) {
+        return 4 + ((numFields + 31) / 32) * 4;
+    }
+
+    /**
+     * It store real value when type is primitive. It store the length and offset of variable-length
+     * part when type is string, map, etc.
+     */
+    public static int calculateFixLengthPartSize(DataType type) {
+        // ordered by type root definition
+        switch (type.getTypeRoot()) {
+            case BOOLEAN:
+            case TINYINT:
+                return 1;
+            case CHAR:
+            case VARCHAR:
+            case BINARY:
+            case VARBINARY:
+            case DECIMAL:
+            case BIGINT:
+            case DOUBLE:
+            case TIMESTAMP_WITHOUT_TIME_ZONE:
+            case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+            case ARRAY:
+            case MAP:
+            case ROW:
+                // long and double are 8 bytes;
+                // otherwise it stores the length and offset of the variable-length part for types
+                // such as is string, map, etc.
+                return 8;
+            case TIMESTAMP_WITH_TIME_ZONE:
+                throw new UnsupportedOperationException();
+            case SMALLINT:
+                return 2;
+            case INTEGER:
+            case FLOAT:
+            case DATE:
+            case TIME_WITHOUT_TIME_ZONE:
+                return 4;
+            default:
+                throw new IllegalArgumentException();
+        }
+    }
+
+    // The number of elements in this array
+    private int size;
+
+    /** The position to start storing array elements. */
+    private int elementOffset;
+
+    public BinaryArrayData() {}
+
+    private void assertIndexIsValid(int ordinal) {
+        assert ordinal >= 0 : "ordinal (" + ordinal + ") should >= 0";
+        assert ordinal < size : "ordinal (" + ordinal + ") should < " + size;
+    }
+
+    private int getElementOffset(int ordinal, int elementSize) {
+        return elementOffset + ordinal * elementSize;
+    }
+
+    @Override
+    public int size() {
+        return size;
+    }
+
+    @Override
+    public void pointTo(MemorySegment[] segments, int offset, int sizeInBytes) {
+        // Read the number of elements from the first 4 bytes.
+        final int size = BinarySegmentUtils.getInt(segments, offset);
+        assert size >= 0 : "size (" + size + ") should >= 0";
+
+        this.size = size;
```

Comment:
set sizeInBytes, offset and segments by super.pointTo

Reply:
done.